### PR TITLE
[SPARK-41532][CONNECT][CLIENT] Add check for operations that involve multiple data frames

### PR DESCRIPTION
**Pull request by Hisoka-X**
### What changes were proposed in this pull request?
Add check for operations that involve multiple data frames,  because spark do not support joining for example two data frames from different Spark Connect Sessions.

### Why are the changes needed?
Spark do not support joining for example two data frames from different Spark Connect Sessions. To avoid exceptions, the client should clearly fail when it tries to construct such a composition.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Add new test

### CODE REVIEW

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/Dataset.scala [hunk 1]
```diff
@@ -560,6 +560,7 @@ class Dataset[T] private[sql] (
   def stat: DataFrameStatFunctions = new DataFrameStatFunctions(sparkSession, plan.getRoot)
 
   private def buildJoin(right: Dataset[_])(f: proto.Join.Builder => Unit): DataFrame = {
+    withDataset(right)
```
Comment (Hisoka-X) [inline connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/Dataset.scala:563]: I will add python implementation later. Just want to make sure my understanding is correct.

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/Dataset.scala [hunk 2]
```diff
@@ -1624,6 +1626,12 @@ class Dataset[T] private[sql] (
     }
   }
 
+  private def withDataset(other: Dataset[_]): Unit = {
```
Comment (hvanhovell) [inline connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/Dataset.scala:1629]: Name it `checkSameSparkSession`?

File: connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/connect/client/util/RemoteSparkSession.scala [hunk 3]
```diff
@@ -49,7 +49,7 @@ import org.apache.spark.util.Utils
 object SparkConnectServerUtils {
 
   // Server port
-  private[connect] val port = ConnectCommon.CONNECT_GRPC_BINDING_PORT + util.Random.nextInt(1000)
+  val port: Int = ConnectCommon.CONNECT_GRPC_BINDING_PORT + util.Random.nextInt(1000)
```
Comment (hvanhovell) [inline connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/connect/client/util/RemoteSparkSession.scala:52]: ??? Why this change?

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/Dataset.scala (hunk 1 above)
Comment (hvanhovell) [reply connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/Dataset.scala:563]: I think the understanding is spot on.

File: connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/connect/client/util/RemoteSparkSession.scala (hunk 3 above)
Comment (Hisoka-X) [reply connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/connect/client/util/RemoteSparkSession.scala:52]: https://github.com/apache/spark/pull/40684/files/f3d1eade9f90517039842fa59815aaf46f0a53b3#diff-7fa161b193c8792c8c0d8dd4bcae3e683ab8553edafa2ae5c13df42b26f612b0R37
just want to use it in `ClientE2ETestSuite` to create another SparkSession. Any better idea?

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/SparkSession.scala [hunk 4]
```diff
@@ -69,6 +70,9 @@ class SparkSession private[sql] (
 
   private[this] val allocator = new RootAllocator()
 
+  // Generate a unique session ID for this session.
+  lazy val sessionId: String = UUID.randomUUID.toString
```
[the remark this answers is not in the document]
Comment (hvanhovell) [reply connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/SparkSession.scala:74]: It would be the same session. If you really want to be sure, just check if the clients are referentially the same.

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/Dataset.scala [hunk 5]
```diff
@@ -1624,6 +1626,12 @@ class Dataset[T] private[sql] (
...
+    assert(
```
[the remark this answers is not in the document]
Comment (Hisoka-X) [reply connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/Dataset.scala]: Done

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/Dataset.scala (hunk 2 above)
Comment (Hisoka-X) [reply connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/Dataset.scala]: Done

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/SparkSession.scala (hunk 4 above)
[the remark this answers is not in the document]
Comment (Hisoka-X) [reply connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/SparkSession.scala:74]: I got your point. Thanks

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/SparkSession.scala [hunk 6]
```diff
@@ -69,6 +69,9 @@ class SparkSession private[sql] (
 
   private[this] val allocator = new RootAllocator()
 
+  // a unique session ID for this session from client.
+  private[sql] lazy val sessionId: String = client.sessionId
```
Comment (hvanhovell) [inline connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/SparkSession.scala:73]: Make this a def? No need for lazy val.

File: connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/connect/client/util/RemoteSparkSession.scala (hunk 3 above)
Comment (HyukjinKwon) [inline connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/connect/client/util/RemoteSparkSession.scala:52]: Can we use a ephemeral port to make sure on the test flakiness? See also `SparkConnectService.localPort`, and https://github.com/apache/spark/blob/c7007b37e15ed1dc7c27e44576aa21f2e9e10e3f/python/pyspark/sql/connect/session.py#L503-L506

Comment (HyukjinKwon) [reply connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/connect/client/util/RemoteSparkSession.scala:52]: Can be done in a separate PR.

BTW, you could use `private[spark] val port: Int` accessor

Comment (HyukjinKwon) [reply connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/connect/client/util/RemoteSparkSession.scala:52]: You could use `private[spark] val port: Int` accessor?

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/SparkSession.scala (hunk 6 above)
Comment (Hisoka-X) [reply connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/SparkSession.scala]: Done

**Review by hvanhovell — APPROVED** [review]
LGTM

**Comment by amaliujia** [comment]
LGTM

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/SparkSession.scala [hunk 7]
```diff
@@ -69,6 +69,9 @@ class SparkSession private[sql] (
 
   private[this] val allocator = new RootAllocator()
 
+  // a unique session ID for this session from client.
+  private[sql] def sessionId: String = client.sessionId
```
Comment (amaliujia) [inline connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/SparkSession.scala:73]: Nit: this to be more accurate should be `user_id+session_id`. However because session_id itself is a random number thus `client.sessionId` should still be able to identify SparkSession uniquely.

Comment (amaliujia) [reply connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/SparkSession.scala:73]: https://github.com/apache/spark/blob/fa6e55bba4d62cdad66e5f425d8a261fe1050134/connector/connect/common/src/main/protobuf/spark/connect/base.proto#L48 this is the user_id of a client.

Comment (Hisoka-X) [reply connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/SparkSession.scala:73]: Good advise, should I change it for this PR? @HyukjinKwon @hvanhovell

Comment (amaliujia) [reply connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/SparkSession.scala:73]: I think current way is fine so I left it as a `Nit`.

**Comment by Hisoka-X** [comment]
@HyukjinKwon @amaliujia @hvanhovell Hi, I add support for python. PTAL again. Thanks!

**Review by amaliujia — APPROVED** [review]
LGTM

**Review by grundprinzip — APPROVED** [review]

**Comment by Hisoka-X** [comment]
@HyukjinKwon @hvanhovell Hi, kindly ask, can merge this PR now?

**Comment by Hisoka-X** [comment]
kindly ping @HyukjinKwon

**Comment by HyukjinKwon** [comment]
I am fine. but I would defer to @hvanhovell to merge.

File: python/pyspark/sql/connect/dataframe.py [hunk 8]
```diff
@@ -249,14 +254,18 @@ def crossJoin(self, other: "DataFrame") -> "DataFrame":
...
         return DataFrame.withPlan(
             plan.Join(left=self._plan, right=other._plan, on=None, how="cross"),
             session=self._session,
         )
 
     crossJoin.__doc__ = PySparkDataFrame.crossJoin.__doc__
 
+    def checkSameSparkSession(self, other: "DataFrame") -> None:
+        if self._session.session_id != other._session.session_id:
+            raise PySparkException("Both Datasets must belong to the same SparkSession")
```
Comment (xinrong-meng) [inline python/pyspark/sql/connect/dataframe.py:267]: I am wondering if a subclass of PySparkException should be used instead of that base class. See https://github.com/apache/spark/blob/master/python/pyspark/errors/exceptions/base.py.

Comment (xinrong-meng) [reply python/pyspark/sql/connect/dataframe.py:267]: Feel free to adjust that in a follow-up though.

Comment (Hisoka-X) [reply python/pyspark/sql/connect/dataframe.py]: Thanks for remind, I add `SessionNotSameException` for this. Please check.

Comment (xinrong-meng) [reply python/pyspark/sql/connect/dataframe.py]: Thanks! It's better to define an error class in python/pyspark/errors/error_classes.py with a template error message. Again, that's not the blocker of merging the PR.

**Review by xinrong-meng — APPROVED** [review]

**Review by hvanhovell — APPROVED** [review]
LGTM
